### PR TITLE
Strip and clean URL field being saved

### PR DESCRIPTION
**Pull request by fysheets**
## EDUCATOR-3346

### CODE REVIEW

File: recommender/recommender.py [hunk 1]
```diff
@@ -1060,4 +1060,24 @@ def strip_and_clean_html_elements(data):
     """
     Clean an HTML elements and return it
     """
-    return bleach.clean(data, strip=True)
+    return bleach.clean(data, tags=[], strip=True)
+
+def strip_and_clean_url(data):
+    """
+    Clean an URL elements and return it
```
Comment (schenedx) [inline recommender/recommender.py]: what is an example of `URL elements`? I think the example you gave for linkify is nice. Can we have something similar here?

File: recommender/recommender.py [hunk 2]
```diff
@@ -586,7 +587,7 @@ def add_resource(self, data, _suffix=''):  # pylint: disable=unused-argument
         for field in self.resource_content_fields:
             result[field] = strip_and_clean_html_elements(data[field])
 
-        resource_id = stem_url(data['url'])
+        resource_id = stem_url(result['url'])
```
Comment (robrap) [inline recommender/recommender.py]: @fysheets: What does stem_url do?  I don't see it being a security issue if you get this wrong, but the `_check_redundant_resource` and `_check_removed_resource` checks below might not do what are intended.  May not be a big deal?

File: recommender/recommender.py [hunk 3]
```diff
@@ -925,11 +926,12 @@ def student_view(self, _context=None):  # pylint: disable=unused-argument
...
-        resources = [{'id': r['id'],
+        resources = [{
+                      'id': strip_and_clean_html_elements(r['id']),
                       'title': strip_and_clean_html_elements(r['title']),
-                      "votes": r['upvotes'] - r['downvotes'],
-                      'url': r['url'],
-                      'description': self._get_onetime_url(r['description']),
+                      "votes": strip_and_clean_html_elements(r['upvotes'] - r['downvotes']),
+                      'url': strip_and_clean_url(r['url']),
+                      'description': strip_and_clean_html_elements(self._get_onetime_url(r['description'])),
```
Comment (robrap) [inline recommender/recommender.py]: What does `_get_onetime_url` mean or do?  Do we have the correct clean method happening here?  Will it work correctly?

File: recommender/recommender.py (hunk 2 above)
Comment (fysheets) [reply recommender/recommender.py]: Looking at the code:

```def stem_url(url):
    """
    Get the base form of url.
    This is not designed for security, just to check common errors/use-cases.
    """
    parsed_url = urlparse(unquote_plus(url))
    return urlunparse(parsed_url._replace(fragment=''))
```

Looks like it does some cleaning as well, but calls out that it's not for security.

File: recommender/recommender.py (hunk 3 above)
Comment (fysheets) [reply recommender/recommender.py]: Looks like `_get_onetime_url` is focusing on `fs` urls:

```    
def _get_onetime_url(self, filename):
        """
        Return one time url for uploaded screenshot

        We benchmarked this as less than 8ms on a sandbox machine.
        """
        if filename.startswith('fs://'):
            return str(self.fs.get_url(filename.replace('fs://', ''), 1000 * 60 * 60 * 10))
        else:
            return filename
```
I don't think that we would want to do `linkify` scrubbing here.

Comment (robrap) [reply recommender/recommender.py]: @fysheets: Would the following make a little more sense?
```
'description': self._get_onetime_url(strip_and_clean_html_elements(r['description'])),
```

File: recommender/recommender.py (hunk 2 above)
Comment (robrap) [reply recommender/recommender.py]: Ok.  We don't have to touch this then.

File: recommender/recommender.py (hunk 3 above)
Comment (fysheets) [reply recommender/recommender.py]: Ah, rereading it, yeah it is more clear that way - I'll make that change quickly and then squash.

**Pull request merged**